content/compute-app/non-functional-requirements/readme.md

```diff
@@ -0,0 +1,156 @@
+
+  The rest of this session focuses on each of the pillars in turn, and talks about the key NFRs in each, using them both as a way to explain the requirements and also as a way to think systematically about non-functional requirements overall.
+
+## Reliability
+
+- **Pillar Concept** - HA/BC/DR - scenarios, availability and recoverability
+
+- **Uptime** is the amount of time that your solution is available and responsive, usually measured as a percentage of time (e.g. 99.9%).
+  - In general, the higher the uptime, the more resiliency you need, and often the more expensive the solution is to run.
+  - We calculate composite SLAs to tell you what Azure is guaranteeing for your solution, although there's more to it than just a single number.
+
+- Recovery from various types of problems is another key element of resiliency.
+- Imagine you've got a solution that you decide to spread across multiple regions in a primary/secondary or hot/cold architecture.
+- Now, a disaster hits your primary region. How long does it take to recover?
+   - In a hot-cold architecture, you might need to stand up some more infrastructure and wait for it to be deployed and configured.
+   - In a warm-standby architecture, you might need to trigger a failover and then wait for DNS records to be updated.
+   - This is your recovery time.
+- **RTO** is the amount of time that you can tolerate being offline in the event of a failure.
+
+- But, suppose your solution deals with a lot of data. During the failover, you might have lost some of this data.
+   - In a hot-cold architecture, you might have last replicated or backed up your data the previous night.
+   - In a warm-standby architecture, you might have asynchronous replication of your data across regions, and maybe it's ~15 minutes out of date.
+   - This is your recovery point.
+- **RPO** is the amoiunt of data that you can tolerate losing in the event of a failure.
+
+- Think about the business continuity in a wider context - e.g. can you switch to manual operations?
+
+- Azure services provide several options for how you architect for different levels of uptime, RTO, and RPO. But these have tradeoffs.
+   - Single-region, single-zone architectures tend to be cheapest, but have the least resiliency. You likely can't support low RTO or RPO.
+   - Multi-region designs are often expensive because you need to deploy multiple separate instances. They can also introduce complexity. And data replication is usually asynchronous, which means you don't automatically get low RPO. But your risk is spread across multiple regions, so they can be resilient and performant, especially when you have customers in multiple regions.
+   - Single-region, multi-zone architectures tend to be the sweet spot for most customers. There is still some cost involved in the redundancy, but you can often support low RPO and RTO requirements.
```

Comment:
Since single-region multi-zone does not protect against DR but provides "better" HA - would it make sense to clarify that? Sometimes customers conflate HA/DR and don't understand the differences (i.e., they need DR but build HA when not required or the opposite).

Reply:
This is a good point, although I'd argue that multi-zone designs _do_ provide some level of DR (against "disasters" like an HVAC outage in a data centre, a localised power outage, etc). But I think your point here is to talk about the tradeoffs and for customers to think about their risk tolerance for different forms of disaster, and if they have a low risk tolerance then they might want to consider multi-region architectures, but with the understanding that this can increase the cost and complexity of the solution.

Reply:
Cool - I've updated to make this a little more explicit too.